src/Surfnet/StepupSelfService/SamlStepupProviderBundle/DependencyInjection/SurfnetStepupSelfServiceSamlStepupProviderExtension.php

```diff
@@ -0,0 +1,206 @@
+        ];
+        $idpAdditional = [
+            'enabled' => false,
+        ];
+
+        $serviceProvider  = array_merge($configuration['service_provider'], $spAdditional, $entityId);
+        $identityProvider = array_merge($idpAdditional, $entityId);
+
+        $hostedDefinition = new Definition('Surfnet\SamlBundle\Entity\HostedEntities', [
+            new Reference('router'),
+            $serviceProvider,
+            $identityProvider
+        ]);
+
+        $hostedDefinition->setPublic(false);
+
+        return $hostedDefinition;
+    }
+
+    /**
+     * @param string           $provider
+     * @param array            $configuration
+     * @param ContainerBuilder $container
+     */
+    private function createRemoteDefinition($provider, array $configuration, ContainerBuilder $container)
+    {
+        $definition    = new Definition('Surfnet\SamlBundle\Entity\IdentityProvider', [
+            [
+                'entityId'        => $configuration['entity_id'],
+                'ssoUrl'          => $configuration['sso_url'],
+                'certificateData' => $configuration['certificate'],
+            ]
+        ]);
+
+        $definition->setPublic(false);
+        $container->setDefinition('gssp.provider.' . $provider . '.remote.idp', $definition);
+    }
+
+    /**
+     * @param string           $provider
+     * @param array            $configuration
+     * @param array            $routes
+     * @param ContainerBuilder $container
+     * @return Definition
+     */
+    private function createMetadataDefinition(
+        $provider,
+        array $configuration,
+        array $routes,
+        ContainerBuilder $container
+    ) {
+        $metadataConfiguration = new Definition('Surfnet\SamlBundle\Metadata\MetadataConfiguration');
+
+        $propertyMap = [
+            'entityIdRoute'          => $this->createRouteConfig($provider, $routes['metadata']),
+            'isSp'                   => true,
+            'assertionConsumerRoute' => $this->createRouteConfig($provider, $routes['consume_assertion']),
+            'isIdP'                  => false,
+//            'ssoRoute'               => $this->createRouteConfig($provider, $routes['sso']),
+            'publicKey'              => $configuration['metadata']['public_key'],
+            'privateKey'             => $configuration['metadata']['private_key'],
+        ];
+
+        $metadataConfiguration->setProperties($propertyMap);
+        $metadataConfiguration->setPublic(false);
+        $container->setDefinition('gssp.provider.' . $provider . 'metadata.configuration', $metadataConfiguration);
+
+        $metadataFactory = new Definition('Surfnet\SamlBundle\Metadata\MetadataFactory', [
+            new Reference('templating'),
+            new Reference('router'),
+            new Reference('surfnet_saml.signing_service'),
+            new Reference('gssp.provider.' . $provider . 'metadata.configuration')
```

Comment:
The more I see this (and yes, I wrote it myself :wink:) the more it seems that the factory should be unaware of which metadata it generates: the configuration should be a parameter of the call to `MetadataFactory::generate()`. Not something that should be fixed here, but if agreed we should create a chore (Tech Debt) and reference that here.